README.md
# FacialRecognition

现在的人脸识别技术已经得到了非常广泛的应用，支付领域、身份验证、美颜相机里都有它的应用。

用Python实现人脸检测。人脸检测是人脸识别的基础。人脸检测的目的是识别出照片里的人脸并定位面部特征点，人脸识别是在人脸检测的基础上进一步告诉你这个人是谁。

## 业务场景


## 实现逻辑

    人脸检测 -> 人脸识别 -> 人脸检测训练 -> 人脸识别训练

**参考来源**

人脸检测部分：[Python爱好者社区](https://mp.weixin.qq.com/s/oDMFoPh6wLYFnTP2rG2AuA) 强哥

人脸识别部分：[CVPy](https://mp.weixin.qq.com/s/3O0m7F712dwnwhpwMXck4Q) 冰不语 

## 依赖库

- Dlib: [http://dlib.net/](http://dlib.net/)
> Dlib是一个跨平台的C++公共库，除了线程支持，网络支持，提供测试以及大量工具等等优点，Dlib还是一个强大的机器学习的C++库，包含了许多机器学习常用的算法。同时支持大量的数值算法如矩阵、大整数、随机数运算等等。Dlib同时还包含了大量的图形模型算法。

**安装所需库**

    pip install numpy
    pip install scipy
    pip install opencv-python
    pip install dlib
    pip install scikit-image

经过查找后得知，Dlib 依赖于 **boots** 和 **CMake**，所以需要先安装这些依赖库，并且配置环境。

> CMake: [https://cmake.org/](https://cmake.org/)  
> boot: [http://www.boost.org/](http://www.boost.org/)  
> **注：**  
> - CMake 3.10.x  
> - boots 1.66.x  
> - Dlib 19.9.x  
> - Dlib-face-landmarks 68 [人脸特征点检测](http://dlib.net/files/shape_predictor_68_face_landmarks.dat.bz2)  
> - dlib-face-recognition-resnet-model [Resnet人脸识别模型](http://dlib.net/files/dlib_face_recognition_resnet_model_v1.dat.bz2)

**scikit-image** 需要 **Microsoft Visual C++ 14.0** 编译器

**boots** 需要 **C++** 系列编译器进行编译 (以下任选其一)

> TDM-GCC: [TDM-GCC](http://tdm-gcc.tdragon.net/)  
> MSVC: [MS Tools](https://www.visualstudio.com/zh-hans/downloads/)

**MSVC** 安装需要7G左右空间，我放弃！！！

如果你是 **windows** 系统，建议使用以下方案

使用方式：文件夹内 pip install xxx.whl

免编译方案：  
下载 [blib-19.8.1](https://pypi.python.org/pypi/dlib/19.8.1 (来自 冰不语)  

**强烈建议 使用 Anaconda**

## 如何运行


## Log

**问题记录：** [Error Log](./ErrorLog.md)

**开发记录：** [Change Log](./Changelog.md)

Changelog.md
# Change log

## 2018年1月24日

人脸检测部分已经完成了，然后根据自己的理解，将代码进行了简单的拆分成可复用调用的方法

- 人脸检测方法：**facial_regional_marker** @params imagePath 需要识别的图片路径

> 可更新部分：接受参数不限于项目路径，即可读取多种来源路径（file path|url|相机|摄像头|……）

---

## 2018年1月25日

人脸识别部分完成，原来案例的代码也拆分出来了不同方法

**关于识别部分说明：** 对于面部识别的耗时稍长，精准度并不是很高，原案例的代码我并没有跑过，所以没确认问题是源案例的还是我更改代码后的。

- 面部模型采集方法：**facial_model_acquisition** 默认读取 **faces** 文件夹下的图片 
- 人脸检测方法：**facial_feature** @params imagePath 需要识别的图片路径

> 可更新部分：目前还需手动添加面部模型对应名称，读取路径可独立成一个方法，返回 **image** 对象

---

## 2018年1月26日

关于面部识别耗时较长的问题进行简单的梳理。

开始对代码进行优化，在简单的整理代码后发现 **可能** 存在性能问题的地方进行改造。

将 **可能** 部分优化完后后续再进行性能测试来具体发生在那个阶段的问题。

1. 将引入模型进行提取为独立的方法在运行时进行初始化，不再是方法内调用。
2. 遍历文件优化方法，将 **glob** 替换成 **os.listdir**。
3. 将 **detector** 配置第二参数取消，直接使用原图可以提升 约200ms 左右的处理时间。

> 关于正面人脸检测器 **detector**  
> **detector(image, 1), detector = dlib.get_frontal_face_detector()**  
> 单张耗时 150~450ms，未尝试过大体积文件 size > 1024kb

> 关于人脸特征描述 **descriptor**  
> **descriptor.compute_face_descriptor(image, shape)**  
> 单张耗时 > 500ms，未尝试过大体积文件 size > 1024kb

> 关于遍历优化，测试了数据样本为 10, 130, 310, 1330, 2660, 5320个文件的情况。  
> 测试方法为 **glob.glob**，**os.listdir**，**os.walk** 三种

**测试代码：**

```python
from time import clock
dirname = './imgs'
def testGLOB():
    start_time = clock()
    for file in glob.glob(dirname + '/*'):
        head, tail = os.path.split(file)
    end_time = clock()
    print("method glob costs time is : " , str(end_time - start_time))

def testListdir():
    start_time = clock()
    for file in os.listdir(dirname):
        os.path.join(dirname, file)
    end_time = clock()
    print("method list costs time is : " , str(end_time - start_time))

def testWalk():
    start_time = clock()
    for(dirname, subshere, fileshere) in os.walk('.'):
        for fname in fileshere:
            os.path.join(dirname, fname)
    end_time = clock()
    print("method walk costs time is : " , str(end_time - start_time))
```

**结论：**

> **os.listdir** 在文件越多的情况下优势越明显  
> **os.listdir** > **glob.glob** > **os.walk**

---

## 2018年1月29日

动态追踪学习实现

参考 [http://python.jobbole.com/81593/](http://python.jobbole.com/81593/)

**原理：** 主要是利用了摄像头读取到的 **第一帧(first frame)** 作为背景模型，以及对读取到的每一帧都做预处理，先转 **灰度图**，再进行 **高斯滤波**，然后会对这些帧进行计算与背景的差异值，从而得到 **分差图（different map）**，这样就可以得到和捕获动态物体的轮廓了，最后再输出出来即可。

```python
# Camera object.
camera = cv2.VideoCapture(0)

# Read the camera frame, it's loop.
while Ture:
    grabbed, frame = camera.read()

    ……
    # Color revised
    gray = cv2.cvtColor(frame, cv2.COLOR_BGR2GRAY)
    # GaussianBlur
    gray = cv2.GaussianBlur(gray, (21, 21), 0)

    ……

    # different map
    frameDelta = cv2.absdiff(firstFrame, gray)
    # threshold to binary
    thresh = cv2.threshold(frameDelta, 25, 255, cv2.THRESH_BINARY)[1]
    # Morphological dilation.  
    thresh = cv2.dilate(thresh, es, iterations = 2)
    _, contours, _ = cv2.findContours(thresh.copy(), cv2.RETR_EXTERNAL, cv2.CHAIN_APPROX_SIMPLE)

```

test_method.py
import os, dlib, cv2
from time import clock
from skimage import io

dirname = './imgs'

# test foreach
def testGLOB():
    start_time = clock()
    for file in glob.glob(dirname + '/*'):
        head, tail = os.path.split(file)
        # print(head, tail, '=>', ('C:\\Other\\' + tail))
    end_time = clock()
    print("method glob costs time is : " , str(end_time - start_time))

def testListdir():
    start_time = clock()
    for file in os.listdir(dirname):
        os.path.join(dirname, file)
        # print(dirname, file, '=>', )
    end_time = clock()
    print("method list costs time is : " , str(end_time - start_time))

def testWalk():
    start_time = clock()
    for(dirname, subshere, fileshere) in os.walk('.'):
        # print('[' + dirname + ']')
        for fname in fileshere:
            os.path.join(dirname, fname)
            # print()
    end_time = clock()
    print("method walk costs time is : " , str(end_time - start_time))

# testGLOB()
# testListdir()
# testWalk()

# test read image
def vc2imread():
    start_time = clock()

    for file in os.listdir(dirname):
        image = cv2.imread(os.path.join(dirname, file))

    end_time = clock()
    print("CV2.imread time is : " , str(end_time - start_time))

def ioimread():
    start_time = clock()

    for file in os.listdir(dirname):
        image = io.imread(os.path.join(dirname, file))

    end_time = clock()
    print("io.imread time is : " , str(end_time - start_time))

vc2imread()
ioimread()

FacialRecognition.py
# -*- coding: UTF-8 -*-
import sys, os, dlib, cv2
import numpy as np

shape_predictor_68_face_landmarks = "./DataModel/shape_predictor_68_face_landmarks.dat"
face_recognition_resnet_model = "./DataModel/dlib_face_recognition_resnet_model_v1.dat"
face_folder_path = "./faces"
face_model_list = []

detector = None
predictor = None
descriptor = None

def init_facial_model():
    global detector, predictor, descriptor
    detector = dlib.get_frontal_face_detector()
    predictor = dlib.shape_predictor(shape_predictor_68_face_landmarks)
    descriptor = dlib.face_recognition_model_v1(face_recognition_resnet_model)

# Facial regional marker to Rect boundary information
def facialRM_to_rect(rect):
    x = rect.left()
    y = rect.top()
    w = rect.right() - x
    h = rect.bottom() - y
    return (x, y, w, h)

# Facial feature marker(shape) to Numpy array
def shape_to_np(shape, dtype = "int"):
    coords = np.zeros((68, 2), dtype = dtype)
    for i in range(0, 68):
        coords[i] = (shape.part(i).x, shape.part(i).y)
    return coords

# Resize image file
def resize(image, width = 1080):
    r = 0
    dim = None

    if image.shape[1] > width:
        r = width * 1.0 / image.shape[1]
        dim = (width, int(image.shape[0] * r))
    else:
        dim = (image.shape[1], image.shape[0])
    
    resized = cv2.resize(image, dim, interpolation = cv2.INTER_AREA)    
    return resized

# Facial regional marker to image file
def facial_regional_marker(imagePath):

    image = cv2.imread(imagePath)
    image = resize(image, width = 1080)
    gray = cv2.cvtColor(image, cv2.COLOR_BGR2GRAY)
    rects = detector(gray)

    for (i, rect) in enumerate(rects):
        shape_predictor = predictor(gray, rect)
        shape = shape_to_np(shape_predictor)

        (x, y, w, h) = facialRM_to_rect(rect)
        cv2.rectangle(image, (x, y), (x + w, y + h), (0, 255, 0), 2)

        cv2.putText(image,
            "Face #{}".format(i + 1),
            (x - 10, y - 10),
            cv2.FONT_HERSHEY_SIMPLEX,
            0.5, (0, 255, 0), 2)
            
        for (x, y) in shape:
            cv2.circle(image, (x, y), 2, (0, 0, 255), -1)

    cv2.imshow("Output", image)
    cv2.waitKey(0)

# Facial_model_acquisition
def facial_model_acquisition():

    for file in os.listdir(face_folder_path):
        image = cv2.imread(os.path.join(face_folder_path, file))

        # 单张耗时 150~450ms，未尝试过大体积文件 size > 1024kb
        # rects = detector(image, 1)
        rects = detector(image)

        for i, rect in enumerate(rects):
            shape = predictor(image, rect)

            # 计算人脸特征描述部分，单张耗时 > 500ms，未尝试过大体积文件 size > 1024kb
            face_descriptor = descriptor.compute_face_descriptor(image, shape)

            features_coordinate = np.array(face_descriptor)
            face_model_list.append(features_coordinate)

def facial_feature(imagePath):

    image = cv2.imread(imagePath)
    image = resize(image, width = 1080)
    gray = cv2.cvtColor(image, cv2.COLOR_BGR2GRAY)
    rects = detector(image)

    x = 0
    y = 0
    dist = []

    for (i, rect) in enumerate(rects):
        shape = predictor(gray, rect)
        x = rect.left()
        y = rect.top()

        face_descriptor = descriptor.compute_face_descriptor(image, shape)

        features_coordinate = np.array(face_descriptor)

        for i in face_model_list:
            np_ln = np.linal
    for (i, rect) in enumerate(rects):
        shape = predictor(gray, rect)
        x = rect.left()
        y = rect.top()

        face_descriptor = descriptor.compute_face_descriptor(image, shape)

        features_coordinate = np.array(face_descriptor)

        for i in face_model_list:
            np_ln = np.linalg.norm(i - features_coordinate)
            dist.append(np_ln)

# '邓丽君', '邓丽君', '邓丽君', '邓丽君', '邓丽君', 'jay', 'jay', 'jay', 'jay',
    candidate = ['denglijun',  'jay']
    temp_dict = dict(zip(candidate, dist))

    dict_sorted = sorted(temp_dict.items(), key = lambda d:d[1])

    cv2.putText(image,
        "{}".format(dict_sorted[0][0]),
        (x - 10, y - 10),
        cv2.FONT_HERSHEY_SIMPLEX,
        0.5, (0, 255, 0), 2)

    cv2.imshow("Output", image)
    cv2.waitKey(0)

def start():
    from time import clock
    start_time = clock()

    init_facial_model()
    # if len(sys.argv) < 2:
    #     print("Usage: %s <image file>" % sys.argv[0])
    #     sys.exit(1)
    facial_model_acquisition()
    # facial_regional_marker("./imgs/face.jpg")
    # facial_feature("./imgs/denglijun-4.jpg")
    facial_feature("./imgs/jay-3.jpg")

    end_time = clock()
    print("method walk costs time is : " , str(end_time - start_time))

start()

DynamicTracking.py
import cv2, time, datetime
import numpy as np

def video_work():
    # Camera object, Get it first camera.s
    camera = cv2.VideoCapture(0)
    # init background model, it's camera first frame.
    firstFrame = None

    time.sleep(0.25)

    size = (int(camera.get(cv2.CAP_PROP_FRAME_WIDTH)),
            int(camera.get(cv2.CAP_PROP_FRAME_HEIGHT)))
    print('size:'+repr(size))

    es = cv2.getStructuringElement(cv2.MORPH_ELLIPSE, (9, 4))
    kernel = np.ones((5, 5), np.uint8)

    while True:
        grabbed, frame = camera.read()

        # If camera can't grabbed frame, exit this app.
        if not grabbed:
            break

        # Color revised.
        gray = cv2.cvtColor(frame, cv2.COLOR_BGR2GRAY)
        # GaussianBlur.
        gray = cv2.GaussianBlur(gray, (21, 21), 0)

        if firstFrame is None:
            firstFrame = gray
            continue

        # different map
        frameDelta = cv2.absdiff(firstFrame, gray)
        # threshold to binary
        thresh = cv2.threshold(frameDelta, 25, 255, cv2.THRESH_BINARY)[1]
        # Morphological dilation. 
        thresh = cv2.dilate(thresh, es, iterations = 2)
        _, contours, _ = cv2.findContours(thresh.copy(), cv2.RETR_EXTERNAL, cv2.CHAIN_APPROX_SIMPLE)

        for c in contours:
            # if the contour is too small, ignore it
            if cv2.contourArea(c) < 1500:
                continue

            # compute the bounding box for the contour, draw it on the frame,
            # and update the text
            # 计算轮廓的边界框，在当前帧中画出该框
            (x, y, w, h) = cv2.boundingRect(c)
            cv2.rectangle(frame, (x, y), (x + w, y + h), (0, 255, 0), 2)
        
        cv2.putText(frame, "Room Status: {}".format("111"), (10, 20),
        cv2.FONT_HERSHEY_SIMPLEX, 0.5, (0, 0, 255), 2)
        cv2.putText(frame, datetime.datetime.now().strftime("%A %d %B %Y %I:%M:%S%p"),
            (10, frame.shape[0] - 10), cv2.FONT_HERSHEY_SIMPLEX, 0.35, (0, 0, 255), 1)
    
        # show it camera current frame.
        cv2.imshow("Security Feed", frame)
        cv2.imshow("Thresh", thresh)
        cv2.imshow("Frame Delta", frameDelta)
        
        key = cv2.waitKey(1) & 0xFF
    
        # if key down is 'q', break loop.
        if key == ord("q"):
            break
    
    # clear camera object and close all the open windows.
    camera.release()
    cv2.destroyAllWindows()

video_work()
